feat(chore): add vectorStore class

Itzaprado opened the pull request:
[ ] Adding create and delete methods
[ ] Adding indexes: StrategyMixin and DistanceStrategy 
[ ] Adding EOF to utils file

Itzaprado committed:
feat(chore): add vectorStore class
[ ] Adding create and delete methods
[ ] Adding indexes: StrategyMixin and DistanceStrategy
[ ] Adding eof to utils file

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -0,0 +1,202 @@
+ * @returns PostgresVectorStore instance.
+ */
+
+class PostgresVectorStore {
```
This should extend the VectorStore class https://github.com/langchain-ai/langchainjs/blob/main/langchain-core/src/vectorstores.ts#L553

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -0,0 +1,202 @@
+class PostgresVectorStore {
+  private static _createKey = Symbol();
+  engine: PostgresEngine;
+  embeddingService: Embeddings;
```
To match the interface this needs to be "embeddings"

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -0,0 +1,202 @@
+      )
+  }
+
+  async delete(ids?: any[] | undefined): Promise<Boolean> {
```
To match the interface this needs to take in a params dictionary and return Promose<void> https://github.com/langchain-ai/langchainjs/blob/5bdbb755bdbb05c3f2bca30c8bd3c64b8295b02a/langchain-core/src/vectorstores.ts#L439

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/tests/vectorStore.test.ts:
```diff
@@ -0,0 +1,132 @@
+      max: 5
+    }
+
+    PEInstance = await PostgresEngine.from_engine_args(url, poolConfig);
```
nit: I would prefer if we use the from_instance method for better security.

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/tests/vectorStore.test.ts:
```diff
@@ -0,0 +1,132 @@
+import { test } from "@jest/globals";
+import PostgresVectorStore, { PostgresVectorStoreArgs } from "../vectorStore.js";
+import PostgresEngine, { Column, VectorStoreTableArgs } from "../engine.js";
+import { SyntheticEmbeddings } from "@langchain/core/utils/testing";
```
How are you installing this? I am trying to run the test locally but not finding this module

Itzaprado committed:
refactor: PostgresVectorStore extends VectorStore
[ ] Change embeddings name variable
[ ] Add missing methods declaration
[ ] Change engine creation to use from_instance

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -0,0 +1,223 @@
+ * @param {string} schema_name  Database schema name of the table. Defaults to "public".
+ * @param {string} content_column Column that represent a Document's page_content. Defaults to "content".
+ * @param {string} embedding_column Column for embedding vectors. The embedding is generated from the document value. Defaults to "embedding".
+ * @param {Array<string} metadata_columns Column(s) that represent a document's metadata.
```
Missing ">"

averikitsch replied:
```suggestion
 * @param {Array<string>} metadata_columns Column(s) that represent a document's metadata.
```

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -0,0 +1,223 @@
+  lambdaMult: Number;
+  indexQueryOptions: QueryOptions;
+
+  constructor(
```
The constructor needs to match the interface (https://github.com/langchain-ai/langchainjs/blob/3c53dcd561705ce3057d80271690d4cae4c236d4/langchain-core/src/vectorstores.ts#L581) therefore this probably needs to be `constructor(embeddings: EmbeddingsInterface, config) {`

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -0,0 +1,223 @@
+
+  _vectorstoreType(): string { // TODO: implement methods
+    //throw new Error("Method not implemented.");
+    return ""
```
```suggestion
    return "cloudsqlpostgresql"
```

Itzaprado committed:
refactor: match constructor with interface
[ ] return correct string on _vectorstoreType
[ ] Add missing characters

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -0,0 +1,190 @@
+   * @param _params - Flexible key-value pairs defining conditions for document deletion.
+   * @returns A promise that resolves once the deletion is complete.
+   */
+  async delete(_params?: Record<string, any>): Promise<void> { // TODO: test this method
```
Let's just support deletion by ids
```suggestion
  async delete(params: {ids?: string[]}): Promise<void> { // TODO: test this method
```

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -0,0 +1,190 @@
+        metadataJsonColumn,
+        distanceStrategy,
+        k,
+        fetch_k,
```
Let's use `fetchK`

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -0,0 +1,190 @@
+   */
+   constructor(embeddings: EmbeddingsInterface, dbConfig: Record<string, any>) {
+    super(embeddings, dbConfig);
+    this.embeddings = embeddings;
```
Correct me if I'm wrong, but don't you need to set "this" for all arguments of the config like this.schemaName?

Itzaprado replied:
Yes, you're right I forgot to update it. Thanks

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -0,0 +1,190 @@
+  indexQueryOptions?: QueryOptions // optional
+}
+
+class PostgresVectorStore extends VectorStore {
```
We also need to set FilterType. We can copy what PGVectorStore has https://github.com/langchain-ai/langchainjs/blob/ded94d5bc77d53a78865d3a7645dbdfeac2735da/libs/langchain-community/src/vectorstores/pgvector.ts#L196

Itzaprado committed:
refactor:
[ ] add FilterType declaration
[ ] change fetch_k to fetchK
[ ] add missing references
[ ] change delete params to support deletion by ids

averikitsch commented on libs/langchain-google-cloud-sql-pg/src/vectorStore.ts:
```diff
@@ -0,0 +1,223 @@
+ * @param {string} content_column Column that represent a Document's page_content. Defaults to "content".
+ * @param {string} embedding_column Column for embedding vectors. The embedding is generated from the document value. Defaults to "embedding".
+ * @param {Array<string} metadata_columns Column(s) that represent a document's metadata.
+ * @param {Array<string} ignore_metadata_columns Optional - Column(s) to ignore in pre-existing tables for a document's metadata. Can not be used with metadata_columns.
```
```suggestion
 * @param {Array<string>} ignore_metadata_columns Optional - Column(s) to ignore in pre-existing tables for a document's metadata. Can not be used with metadata_columns.
```
